### ISSUE TITLE
Add battery "Maximum Discharging Power" as a configurable NumberEntity

### ISSUE THREAD

**Issue by wlcrs** [wlcrs/huawei_solar#15]
I’ve been working on exposing some parameters as configurable entities in Home Assistant. My primary use case would be to make it possible to set the “Maximum Discharge Power” for my LUNA2000 battery from Home Assistant. This would allow me to adjust the power output of the battery dynamically based on other inputs.

Example use case: in Belgium, you are taxed on peak usage, starting when you draw more than 2.5kW from the grid for a certain time. I would like to create an automation that sets the “Maximum Discharging Power” for my battery to “Current Grid Power Usage (in W) - 2500W”.

However, while the register is mentioned as being read/writeable in the Modbus Interface Definitions document, I’m receiving a “Permission denied” error (0x80) when trying to write to it.

![image](https://user-images.githubusercontent.com/2150060/149144401-dd4e6dc5-ac96-4550-92ef-29d2bbb6940f.png)

Support is letting me know that the function is not available at the moment :frowning: :

![image](https://user-images.githubusercontent.com/2150060/149144371-8c5a4396-b584-4238-8e70-e79bb21bfde2.png)

For everyone who would also be interested in using this function: Please send a mail to Huawei support to ask for this functionality! The more requests are made, the more hope that the function will ever be enabled.

An example mail:

> Dear,
> 
> I want to use the Modbus-interface on my inverter to dynamically set the maximum discharging power of my LUNA2000 battery.
> The Modbus Interface Definitions state that register 47077 is RW. However, when I try to write a value to this register, I get the error code 0x80, which means “No permission”?
>
> My inverter SN: HVxxx> 
> Battery SN: BTyyy

**Comment by inigoml** [comment]
Hi @wlcrs 

It's a very interesting feature. Not only discharge power, but also charge/discharge mode (maximum autoconsumption on sunny days, TOU in cloudy days or winter for charging in low price hours), definable TOU charge&discharge hours or battery capacity grid cutoff percentage based on energy generation forecast in order to establish a low charge level for sunny days or higher charge level for cloudy days.

Will write to EU Support also.

**Comment by wlcrs** [comment]
Hi,

An update from my side. I pressed Huawei Europe for some more information,
and they informed me that an SDongle is needed in order to be able to write
to the registers to my SUN2000-3K-L1. The modbus interface which is exposed
by my inverter directly is not authorized apparently.

I'm now in contact with my installer to figure out the next steps. But it
takes some time.

I'm currently finishing a big rewrite/update from Emilv2's library that
does the actual modbus communication. Once that is done, I'll submit an
official integration for HA (based on the work done in this repo).
Initially this integration will only read from the inverter, but
write-support is still on the roadmap after that.

**Comment by inigoml** [comment]
Since you can alter inverter behavior from SUN2000 application, for sure we
can alter inverter parameters from SUN2000-XXXXX network. Not sure if
SUN2000 app uses modbus protocol or other Huawei custom one for modify
parameters but this way for sure exists.
Another way would be using REST API from Fusionsolar web. But you need an
account with installer privileges at website.

**Comment by inigoml** [comment]
Response from Huawei support:
![image](https://user-images.githubusercontent.com/7486708/151232485-c00aee0b-ce29-4490-becc-77cd29e8dec9.png)

**Comment by inigoml** [comment]
They also sent me this document:
![image](https://user-images.githubusercontent.com/7486708/151232594-f2d5a749-1c8d-4866-91c4-439fe99c8e4b.png)

**Comment by inigoml** [comment]
@wlcrs, do you have some code snippet in order to test with other modbus commands?
I'm trying to modify this:

![image](https://user-images.githubusercontent.com/7486708/151232888-b9577aec-cdf5-4340-b49e-41a22710e6a7.png)

And also this:
![image](https://user-images.githubusercontent.com/7486708/151233162-f544088b-6a05-4888-a360-762aba7e5d58.png)

**Comment by inigoml** [comment]
@wlcrs, do you have some code snippet in order to test with other modbus commands?
I'm trying to modify this:

![image](https://user-images.githubusercontent.com/7486708/151232888-b9577aec-cdf5-4340-b49e-41a22710e6a7.png)

And also this:

**Comment by wlcrs** [comment]
I have a branch with the poc for writing to registers. Have a look at this:
https://gitlab.com/wlcrs/huawei-solar/-/blob/write-support/temp.py

**Comment by wlcrs** [comment]
I’ve worked out how to do the login on the inverter in the same way the FusionSolar app does. This enables write-support for some configuration options, all related to having a battery (maximum charge power, maximum discharge power, charging from the grid, …).

![image](https://user-images.githubusercontent.com/2150060/151712423-1c70b19a-2579-426b-a5b3-a659e34b6c2e.png)

You can test it by using the code from the `login-and-write` branch. Note that it requires the installer account to have the default password 00000a for now.

I haven’t been able to test this new functionality for long-term stability yet. Please report any unexpected behavior to me.

**Comment by inigoml** [comment]
Wow... you are my hero!

**Comment by inigoml** [comment]
Working but... also not writable. 
![image](https://user-images.githubusercontent.com/7486708/151717339-d3a49c18-8f21-41b2-8ad7-2f9510fff9a6.png)

**Comment by wlcrs** [comment]
That will be a mistake from my end, not an error returned by the inverter.
I'll fix it tomorrow.

In the meantime: I also implemented a few services which allow you to start
a forcible charge/discharge for a certain duration of time, or until a
specific SoC is reached. Check them in the services tab of the developers
tools

**Comment by inigoml** [comment]
I'm moving right now from emilv2 huawei_solar to your branch. I had to remove all fixes and workarounds in order to make Energy Stats working. Yours is pretty clean!

**Comment by inigoml** [comment]
With this branch you could eventually eliminate the need of install SUN2000 commissioning app!

![image](https://user-images.githubusercontent.com/7486708/151719700-5f758741-314e-490f-aff0-39b2225c451c.png)

**Comment by wlcrs** [comment]
Just tested those services again myself, but it seems like I have broken it
just before I finished working on them. Will look into it next week.

**Comment by wlcrs** [comment]
I've fixed a few register-settings, and also added a "stop forcible charge" service. Can you do a 'git pull' and try again?

I've tested the services, and they seem to be working well now. 

An observation: the forcible charge commands will not charge from grid if the solar panels are producing anything at all. Not surprising that the inverter can't do both at once. Instead, all production will be diverted into the battery until the target is reached.

**Comment by inigoml** [comment]
It appears is working right now (no error). Will check later via SUN2000 to ensure values are also set up at inverter.
This is a really nice feature!

**Comment by JacoboSanchez** [comment]
I am posting here as it is somehow related with the topic. I am testing the write feature to figure out how to automate a battery load at night with lower prices and I am getting sometimes permission denied errors. My guess is that previously authentication is somehow lost and the inverter requires to authenticate again. Did anyone experienced this? 

FYI: I am still using the network interface on SP109 firmware (not SDongle and not connected to the built-in access point)

**Comment by inigoml** [comment]
I've tested this feature with these firmware versions:

SPC114 -> Modbus connection to inverter IP in my wifi network (port 502), and modbus connection to inverter IP in inverter wifi network (SUN2000-XXXXXXX), port 502.
SPC117 -> Modbus connection to inverter IP in inverter wifi network (SUN2000-XXXXXXX), port 6607
SPC119 -> Modbus connection to inverter IP in inverter wifi network (SUN2000-XXXXXXX)., port 6607

Feel free to upgrade to SPC114 but not go further. In general, SPC114 is the best one due to no wifi network issues and modbus opened in regular wifi (SPC117 and SPC119 has "once a day" disconnection in my wifi network and I have to enter via FusionSolar commissioning app in order to reconnect wifi).

**Comment by wlcrs** [comment]
Can you increase your logging (cfr. bottom of the FAQ in the README), wait till the issue occurs, and post (a link to) your home-assistant.log in a new issue?

As I do frequent restarts of my own HA-instance (because I need to test small fixes here and there), I haven't been able to test long-running sessions myself.

I want to know if the underlying logic is sound before I add retry-mechanisms to login again.

**Comment by inigoml** [comment]
> I want to know if the underlying logic is sound before I add retry-mechanisms to login again.

Regarding retry, due to frequent loss of connectivity with inverter (at least for those models with internal wifi, not sure with FE dongle), sometimes automations do not work due to reconnection is read-only. I don't know why. Only way to fix this is restarting HomeAssistant.

**Comment by JacoboSanchez** [comment]
> Can you increase your logging (cfr. bottom of the FAQ in the README), wait till the issue occurs, and post (a link to) your home-assistant.log in a new issue?
> 
> As I do frequent restarts of my own HA-instance (because I need to test small fixes here and there), I haven't been able to test long-running sessions myself.
> 
> I want to know if the underlying logic is sound before I add retry-mechanisms to login again.

I have activated the log (pymodbus included) so I will create another Issue when reproduced, thanks.

**Comment by JacoboSanchez** [comment]
Well it refuses to fail now... maybe I must wait until tomorrow. In the future I plan to install a modbus  rtu-tcp bridge to be able to connect without using Huawei's modbus tcp implementation. This way I should have write permissions without login... At that point I will test if that works.
I already have one inverter using the bridge (Elfin EW11) but it has no battery so I have no write options to test

**Comment by wlcrs** [comment]
The integration should already support this scenario. When you connect over your rtu-tcp bridge, and check the "Advanced: enable parameter configuration" checkbox, it should skip the screen in which it asks for the installer username/password. 

I would greatly appreciate it if you let me know if that works as expected or not.

You could already test it with your current setup. Checking that checkbox will just try to write the timezone-register. It should not produce any errors. It will not produce any configurable entities of course.

**Comment by JacoboSanchez** [comment]
![image](https://user-images.githubusercontent.com/11249890/154118943-13b7d233-f755-48d0-a2ea-0d4517850cea.png)
![image](https://user-images.githubusercontent.com/11249890/154118994-945a136e-9980-490e-900e-90bddccabf5c.png)
And... maybe this is the proof?
![image](https://user-images.githubusercontent.com/11249890/154120793-3d9c2318-432d-49ae-abfc-b7d37186a86a.png)

**Comment by wlcrs** [comment]
The pymodbus logs are from a heartbeat of your other inverter. But it's very good news that using your rtu-tcp bridge worked so smoothly!

Are your inverters not connected to each other via the RS485_2 interface? You should be able to read from both of them by connecting to the "master" inverter and mentioning all slave ids (comma separated) in the first config screen.

**Comment by denislooby** [comment]
It's interesting that the docs say the range for the Grid Charge Cutoff SOC is 30-70 but using fusion solar and the installer account directly connected I can use higher cutoff than that.
I wonder if it varies by inverter or battery, LG vs Luna, it's a Luna in my case.

**Comment by wlcrs** [comment]
@denislooby . Yeah, I first implemented it as per the modbus register definitions, but in #25 this was changed to reflect the reality of what FusionSolar allows.

I'm closing this issue as this feature has been adequately tested by various people now
